File: Question_01_10/Answers_Py/answer_02.py
##########################################
#    Question 02 : Grayscale(灰度化)
##########################################
import cv2 as cv
import numpy as np

# Target : gray scale
def BGR2GRAY(img):
    b = img[:, :, 0].copy()
    g = img[:, :, 1].copy()
    r = img[:, :, 2].copy()
    # Gray scale
    out = 0.2126 * r + 0.7152 * g + 0.0722 * b
    out = out.astype(np.uint8)
    return out

# Read Image
img = cv.imread("./imori.jpg")
# Show & Save origin image
cv.imshow("Origin", img)
cv.imwrite("Origin_02.jpg", img)
# Gray scale
img = BGR2GRAY(img)
# Save result
cv.imwrite("Output_02.jpg", img)
cv.imshow("Result", img)
cv.waitKey(0)
cv.destroyAllWindows()

File: Question_01_10/Answers_Cpp/answer_02.cpp
/*
 * Question 02 : GrayScale(灰度化) 
 * Created by Harrytsz 2021-1-2
 */
#include <iostream>  
#include <opencv2/core/core.hpp>  
#include <opencv2/highgui/highgui.hpp>  

using namespace cv;

Mat BGR2GRAY(Mat img) {
	// get height and width
	int height = img.rows;
	int width = img.cols;
	// prepare output
	Mat out = Mat::zeros(height, width, CV_8UC1);
	for (int y = 0; y < height; y++) {
		for (int x = 0; x < width; x++) {
			// BGR -> GRAY
			out.at<uchar>(y, x) = 0.2126 * (float)img.at<Vec3b>(y, x)[2] \
				+ 0.7152 * (float)img.at<Vec3b>(y, x)[1] \
				+ 0.0722 * (float)img.at<Vec3b>(y, x)[0];
		}
	}
	return out;
}


int main()
{
	// Read image    
	Mat img = imread("imori.png");    
	namedWindow("Images");
	// Show Origin image
	imshow("Origin", img);
	imwrite("Origin_02_cpp.jpg", img);
	// Run operation
	Mat out = BGR2GRAY(img);
	// Show output image
	imshow("Output", out);
	imwrite("Output_02_cpp.jpg", out);
	// Esc exit
	waitKey(0);
	destroyAllWindows();
	return 0;
}